### PR TITLE
[WebAssembly] Disable running `wasm-opt` on components

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds a check that disables `wasm-opt` for the `wasm32-wasip2` target because `wasm-opt` doesn't support components at this time. This also fixes a minor issue from #95208 where if `wasm-opt` was disabled then the linker wouldn't run at all.

### CODE REVIEW

File: clang/lib/Driver/ToolChains/WebAssembly.cpp [hunk 1]
```diff
@@ -61,6 +61,10 @@ std::string wasm::Linker::getLinkerPath(const ArgList &Args) const {
   return ToolChain.GetProgramPath(ToolChain.getDefaultLinker());
 }
 
+static bool IsWasip2(const llvm::Triple &TargetTriple) {
+  return TargetTriple.getOSName() == "wasip2";
```
Comment (pavelsavara) [inline clang/lib/Driver/ToolChains/WebAssembly.cpp:65]: I guess WASI preview3 and the future will be the same. Should we invert the test and only check for `wasip1` ?

Comment (alexcrichton) [reply clang/lib/Driver/ToolChains/WebAssembly.cpp]: Sounds reasonable to me yeah 👍

**Review by sbc100 — APPROVED** [review]
Can we write a test for this?

**Comment by alexcrichton** [comment]
Is there a way to "fake" a `wasm-opt` executable? I didn't find any existing tests for the `wasm-opt` integration and it's auto-disabled if `wasm-opt` isn't found on `$PATH`, which doesn't happen by default in the test suite, so unless I can set up `wasm-opt` to show up in `$PATH` as part of the test I'm not sure how to test

**Comment by sbc100** [comment]
> Is there a way to "fake" a `wasm-opt` executable? I didn't find any existing tests for the `wasm-opt` integration and it's auto-disabled if `wasm-opt` isn't found on `$PATH`, which doesn't happen by default in the test suite, so unless I can set up `wasm-opt` to show up in `$PATH` as part of the test I'm not sure how to test

Ah.. I guess if there are no existing tests then maybe this is something that is particularly hard to test.

**Comment by mh4ck-Thales** [comment]
The tests for the use of `wasm-opt` are hard to integrate into the LLVM project as `wasm-opt` is not part of the LLVM toolchain itself (which is the initial use case for the new `--no-wasm-opt` flag: as LLVM is not distributed with `wasm-opt` by default a same compilation command silently resulted of different binaries depending on the presence or not of `wasm-opt`...). I guess maybe one can hack the test suite in some way for this specific case, or maybe a suitable solution could be found to integrate `wasm-opt` into the LLVM distributions.

@alexcrichton would it be relevant to open an issue in the binaryen repo for the support of Wasm components?

**Comment by pavelsavara** [comment]
could shell script named `wasm-opt` on PATH and doing nothing do the trick ?
Or even better it could write it's arguments into some file, so that they could be asserted in the test.

2c

**Comment by mh4ck-Thales** [comment]
A shell script would work. I don't know how easy it is to create a shell script inside the test suite and use it though, and it would be nice if this fake wasm-opt script could be unique across several tests, to ensure the consistency of those.

**Comment by alexcrichton** [comment]
Ah yeah unfortunately I think it still comes back to how to do all this in the test harness. If there are examples of this elsewhere I'd be happy to investigate and copy. I'm also not sure if shell scripts would work for Windows...

@sbc100 if you're ok with this without a test though would you be able to merge for me?

**Comment by alexcrichton** [comment]
> would it be relevant to open an issue in the binaryen repo for the support of Wasm components?

Oh sorry forgot to respond to this as well. I'm not personally part of the binaryen repo but I suspect they would not object to having an issue to track this.

**Review by sbc100 — APPROVED** [review]

**Comment by mh4ck-Thales** [comment]
Seems like bash is present on the GitHub Windows runners. I shouldn't be too hard to create a simple bash script that work both on Windows and Linux. It would add a new dependency for the test suite on Windows however (if bash is not already a dependency for Windows)

**Comment by mh4ck-Thales** [comment]
@alexcrichton sorry I thought that binaryen was under the bytecode alliance and not the webassembly org. I opened the issue here : https://github.com/WebAssembly/binaryen/issues/6728 , and proposed your idea for a solution.

**Comment by alexcrichton** [comment]
@sbc100 oh I can't merge things myself so if you'd be ok merging for me that'd be appreciated

**Comment by alexcrichton** [comment]
ping @sbc100 would you be able to merge this for me?

**Review by sunfishcode — APPROVED** [review]

**Pull request merged**